### PR TITLE
fix(nuxt): handle page rendering on different path

### LINKED ISSUE #20653: Google WebCache issue
### Environment

------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.14.2`
- Nuxt Version:     `3.3.1`
- Nitro Version:    `2.3.1`
- Package Manager:  `pnpm@7.32.2`
- Builder:          `vite`
- User Config:      `experimental`, `runtimeConfig`, `vite`, `hooks`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

https://webcache.googleusercontent.com/search?q=cache:y8j45sa4h-gJ:https://nuxt.com/&cd=15&hl=fr&ct=clnk&gl=fr

https://codesandbox.io/p/github/nuxt/starter/csb-8m6t9s/draft/silly-murdock?file=%2Fnuxt-app%2Fpages%2Findex.vue&selection=%5B%7B%22endColumn%22%3A32%2C%22endLineNumber%22%3A3%2C%22startColumn%22%3A32%2C%22startLineNumber%22%3A3%7D%5D

https://8m6t9s-3000.csb.app/ => 200
cache:https://8m6t9s-3000.csb.app => 404

### Describe the bug

Hi :) ,

We have recently found that we have an issue with google cache with our project.
Seems like the pages are correctly display server side, but when the client takes the hand, a 404 appear.

This is reproducible on nuxt.com page.

image

This issue (https://github.com/vuejs/vue-router/issues/2042) was present on the previous version and solutions was find but doesn't seem to be Nuxt 3 compatible.

Any idea?

### Additional context

_No response_

### Logs

```shell-script
[nuxt] error caught during app initialization Error: Page not found: /search?q
```

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves https://github.com/nuxt/nuxt/issues/20653

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This enables a Nuxt site to render without throwing a 404 if it is displayed on a different route (e.g. from a search engine cached result).

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: test/fixtures/basic/server/routes/proxy.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export default defineEventHandler(async () => {
+  return await $fetch<string>('/')
```
Comment (pi0) [inline test/fixtures/basic/server/routes/proxy.ts:2]: As best practice, we should use `event.$fetch` to preserve provider context

File: packages/nuxt/src/app/nuxt.ts [hunk 2]
```diff
@@ -255,13 +256,13 @@ export function createNuxtApp (options: CreateOptions) {
...
     if (nuxtApp.ssrContext) {
+      // Expose nuxt to the renderContext
       nuxtApp.ssrContext.nuxt = nuxtApp
-    }
-    // Expose payload types
-    if (nuxtApp.ssrContext) {
+      // Expose payload types
       nuxtApp.ssrContext._payloadReducers = {}
+      // Expose current path
+      nuxtApp.payload.path = nuxtApp.ssrContext!.event.path
```
Comment (pi0) [inline packages/nuxt/src/app/nuxt.ts:265]: (Isn't existence of `ssrContext` already asserted in L259?)

**Review by pi0 — APPROVED** [review]
👍🏼

File: test/fixtures/basic/server/routes/proxy.ts (hunk 1 above)
Comment (danielroe) [reply test/fixtures/basic/server/routes/proxy.ts:2]: interesting bug - it does not support a generic 🤔

**Comment by bokub** [comment]
Hi Daniel, I'm sorry to bother you

Do you think this PR could be the root cause of the problem described here? https://github.com/nuxt/nuxt/discussions/32730

Thanks a lot

**Comment by danielroe** [comment]
@bokub indeed. you need to modify `nuxtApp.payload.path` if you want to do something like that.

**Comment by bokub** [comment]
@danielroe Thanks a lot for answering. Any idea where or how I can do that?

I've tried to change `nuxtApp.payload.path` from a plugin, but it's not early enough I think. From what I understand, it should be done before the router is created.

Thanks again

_____

Edit: I found a way to edit the path using an inline script, but **it feels really dirty**, if you have a better idea please don't hesitate to tell me :slightly_smiling_face: 

```ts
// nuxt.config.ts

export default defineNuxtConfig({
  ssr: true,
  app: {
    head: {
      script: [
        {
          textContent: `window.__NUXT__.path = window.location.pathname + window.location.search`,
          tagPosition: 'bodyClose',
        },
      ],
    },
  },
});
```

**Comment by danielroe** [comment]
Here's an example: https://github.com/nuxt/nuxt/issues/21716#issuecomment-1604560916.

**Pull request merged**